### DISCUSSION TITLE
Overload Resolution with Unconstrained Generic Type Testing

### DISCUSSION THREAD

**Discussion by fitdev** [dotnet/csharplang#6269] [Q&A] [answered]
Consider the following code:

```cs
using System;
using System.Diagnostics.CodeAnalysis;

public class Test {
    
    public void M<T>(T a, T b) {
        var _equal = false;
        if(a is IEquatable<T> ae && b is IEquatable<T> be) _equal = ae.Equals(be); // <== Why here object.Equals(object) overload is used instead of IEquatable one despite the cast?
    }
    
}

struct Foo : IEquatable<Foo> {
    
    public int A;

    public bool Equals(Foo other) => A == other.A;

}
```

Question: Why in `ae.Equals(be)` an overload for `object.Equals(object)` is chosen instead of `IEquatable<T>` one despite the cast? Shouldn't the cast take precedence and effectively "constrain" `T` to `IEquatable<T>`?

**Comment by PathogenDavid** [comment, marked as answer]
Because `IEquatable<T>` provides the method `Equals(T)`, not `Equals(IEquatable<T>)`.

You should just use `b` directly instead:

```csharp
if(a is IEquatable<T> ae) _equal = ae.Equals(b);
```

And if you expect an oddball scenario where `a` and `b` have a common `T` but don't both implement `IEquatable<T>` you might extend this to:

```csharp
if(a is IEquatable<T> ae) _equal = ae.Equals(b);
else if(b is IEquatable<T> be) _equal = be.Equals(a);
```

**Reply by fitdev** [reply]
Thanks a lot for a thorough explanation. Silly of me to overlook such a simple solution and not pay close attention to the actual signature of a method.
